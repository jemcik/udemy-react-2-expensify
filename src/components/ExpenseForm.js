import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { Link } from 'react-router-dom';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? String(props.expense.amount / 100) : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }));
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (this.state.description && this.state.amount) {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        } else {
            this.setState(() => ({ error: 'Пожалуйста, добавьте описание и сумму.' }));
        }
    };

    render() {
        return (
            <form className='form' onSubmit={this.onSubmit}>
                {this.state.error && <p className='form-error'>{this.state.error}</p>}
                <input
                    className={`text-input${this.state.error && !this.state.description ? ' error' : ''}`}
                    type='text'
                    placeholder='описание'
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    className={`text-input${this.state.error && !this.state.amount ? ' error' : ''}`}
                    type='number'
                    placeholder='сумма'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className='textarea'
                    placeholder='комментарий (не обязательно)'
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <div>
                    <button className='button'>Сохранить</button>
                    <Link className='button grey' to='/dashboard'>Отмена</Link>
                </div>
            </form>
        );
    }
}
