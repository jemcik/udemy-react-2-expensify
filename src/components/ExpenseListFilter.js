import React from 'react';
import uuid from 'uuid';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilter extends React.Component {
	state = {
		calendarFocused: null
	};

	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));
	};

	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value);
	};

	onFilterChange = (e) => {
		switch (e.target.value) {
			case 'date':
				this.props.sortByDate();
				break;
			case 'amount':
				this.props.sortByAmount();
				break;
		}
	};

	render() {
		return (
			<div>
				<input
					type='text'
					value={this.props.filters.text}
					onChange={this.onTextChange}
				/>
				<select
					value={this.props.filters.sortBy}
					onChange={this.onFilterChange}
				>
					<option value='date'>Date</option>
					<option value='amount'>Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					startDateId={uuid()}
					endDate={this.props.filters.endDate}
					endDateId={uuid()}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => false}
					showClearDates={true}
				/>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({
	filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate)),
	setTextFilter: (value) => dispatch(setTextFilter(value)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
