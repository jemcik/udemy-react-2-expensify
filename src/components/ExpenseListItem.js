import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <Link className='list-item' to={`/edit/${id}`}>
        <div>
            <h3 className='title'>{description}</h3>
            <span className='sub-title'>{moment(createdAt).format('LL')}</span>
        </div>
        <h3 className='data'>{numeral(amount / 100).format('0,0[.]00 $')}</h3>
    </Link>
);

export default ExpenseListItem;
