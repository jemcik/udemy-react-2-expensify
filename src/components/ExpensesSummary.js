import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ selectExpenses }) => (
    <div>
        Viewing {selectExpenses.length} expense{selectExpenses.length > 1 ? 's' : ''} totalling {numeral(expensesTotal(selectExpenses) / 100).format('$0,0.00')}
    </div>
);

const mapStateToProps = (state) => ({
    selectExpenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);
