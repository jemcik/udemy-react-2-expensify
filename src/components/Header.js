import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
    <header className='header'>
        <div className='content-container'>
            <div className='content'>
                <Link className='title' to='/dashboard'>
                    <h1>Учет расходов</h1>
                </Link>
                <button className='button link' onClick={startLogout}>Выйти</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});
export default connect(undefined, mapDispatchToProps)(Header);
