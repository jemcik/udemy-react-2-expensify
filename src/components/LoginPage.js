import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className='box-layout'>
        <div className='box'>
            <h1 className='title'>Расходы</h1>
            <p>Возьми свои расходы под контроль</p>
            <button className='button' onClick={startLogin}>Войти с помощью Google</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
