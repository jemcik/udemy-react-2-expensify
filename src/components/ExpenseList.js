import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';

import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className='content-container'>
        <div className='list-header'>
            <div className='show-for-mobile'>Расходы</div>
            <div className='show-for-desktop'>Расходы</div>
            <div className='show-for-desktop'>Сумма</div>
        </div>
        <div className='list-body'>
            {props.expenses.length === 0 ? (
                <div className='list-item message'>
                    <span>Нет расходов</span>
                </div>
            ) : (
                props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
            )}
        </div>
    </div>
);

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
