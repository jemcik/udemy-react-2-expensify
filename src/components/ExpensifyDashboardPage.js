import React from 'react';
import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';
import ExpenseListFilter from './ExpenseListFilter';

const ExpensifyDashboardPage = () => (
    <div>
        <ExpenseListFilter />
        <ExpensesSummary />
        <ExpenseList />
    </div>
);

export default ExpensifyDashboardPage;
