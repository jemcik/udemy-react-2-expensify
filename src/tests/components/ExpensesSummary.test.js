import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { ExpensesSummary} from '../../components/ExpensesSummary';

test('render ExpensesSummary for 3 expenses', () => {
    const wrapper = shallow(
        <ExpensesSummary
            selectExpenses={expenses}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('render ExpensesSummary for 1 expense', () => {
    const wrapper = shallow(
        <ExpensesSummary
            selectExpenses={[expenses[0]]}
        />
    );
    expect(wrapper).toMatchSnapshot();
});
