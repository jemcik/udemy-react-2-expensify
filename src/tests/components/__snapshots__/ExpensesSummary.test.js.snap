// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpensesSummary for 1 expense 1`] = `
<div>
  Viewing 
  1
   expense
   totalling 
  $1.95
</div>
`;

exports[`render ExpensesSummary for 3 expenses 1`] = `
<div>
  Viewing 
  3
   expense
  s
   totalling 
  $1,141.95
</div>
`;
